### ISSUE TITLE
Where can I set "KernelArch" variable to build seL4?

### ISSUE THREAD

**Issue by mincheolsung** [seL4/seL4#908]
Hi,

I am trying to build seL4 but can't find a location to set "KernelArch". Sorry, I admit that I am not familiar with cmake.
Below was errors I got when `cmake -DCMAKE_TOOLCHAIN_FILE=../kernel/gcc.cmake -G Ninja ..`

CMake Error at CMakeLists.txt:9 (include):
  include could not find load file:

    settings.cmake

-- The C compiler identification is GNU 9.4.0
-- Check for working C compiler: /usr/bin/gcc
-- Check for working C compiler: /usr/bin/gcc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- The CXX compiler identification is GNU 9.4.0
-- Check for working CXX compiler: /usr/bin/g++
-- Check for working CXX compiler: /usr/bin/g++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- The ASM compiler identification is GNU
-- Found assembler: /usr/bin/gcc
CMake Error at kernel/configs/seL4Config.cmake:172 (message):
  Variable 'KernelArch' is not set.
Call Stack (most recent call first):
  kernel/CMakeLists.txt:10 (include)

-- Configuring incomplete, errors occurred!
See also "/home/mincheol/mnt/awesome_system/build/CMakeFiles/CMakeOutput.log".

**Comment by axel-h** [comment]
You also have to set the platform (`-DPLATFORM=...`), a default architecture is derived from this.

**Comment by mincheolsung** [comment]
I still got the same error even with -DPLATFORM=x86 or -DPLATFORM=x86_86

**Comment by axel-h** [comment]
It seems you have a custom build environment. Can you give more details about your directory layout and what CMake files you are invoking? Looks like a `settings.cmake` is also missing, what system are you trying to build?

**Comment by mincheolsung** [comment]
My build tree looks like:

mincheol@ubuntu:~/mnt/awesome_system$ find . -name "CMakeLists.txt"
./projects/seL4_libs/libsel4allocman/CMakeLists.txt
./projects/seL4_libs/libsel4vspace/CMakeLists.txt
./projects/seL4_libs/libsel4utils/CMakeLists.txt
./projects/seL4_libs/libsel4simple-default/CMakeLists.txt
./projects/seL4_libs/libsel4test/CMakeLists.txt
./projects/seL4_libs/libsel4muslcsys/CMakeLists.txt
./projects/seL4_libs/libsel4vka/CMakeLists.txt
./projects/seL4_libs/libsel4sync/CMakeLists.txt
./projects/seL4_libs/libsel4debug/CMakeLists.txt
./projects/seL4_libs/CMakeLists.txt
./projects/seL4_libs/libsel4bench/CMakeLists.txt
./projects/seL4_libs/libsel4serialserver/CMakeLists.txt
./projects/seL4_libs/libsel4platsupport/CMakeLists.txt
./projects/seL4_libs/libsel4simple/CMakeLists.txt
./projects/awesome/CMakeLists.txt
./tools/elfloader-tool/CMakeLists.txt
./kernel/libsel4/CMakeLists.txt
./kernel/CMakeLists.txt
./CMakeLists.txt-> tools/cmake-tool/default-CMakeLists.txt

I followed https://docs.sel4.systems/projects/buildsystem/incorporating.html.

**Comment by Ivan-Velickovic** [comment]
> I still got the same error even with -DPLATFORM=x86 or -DPLATFORM=x86_86

I believe it should be `-DPLATFORM=ia32` for 32-bit and `-DPLATFORM=x86_64` for 64-bit. Do either of those work?

**Comment by mincheolsung** [comment]
> -DPLATFORM=x86_64

I tried -DPLATFORM=x86_64, but got the same error. Also, as @axel-h pointed out, I could not find settings.cmake.

**Comment by mincheolsung** [comment]
Guys,

I added
set(KernelArch KernelArchX86) in seL4/config.cmake
set(KernelArch KernelArchX86) in seL4/gcc.cmake
set(KernelWordSize 64) in seL4/gcc.cmake

but got this error below:
mincheol@ubuntu:~/mnt2/awesome_system/build$ cmake -DCMAKE_TOOLCHAIN_FILE=../kernel/gcc.cmake -G Ninja ..
CMake Error at CMakeLists.txt:9 (include):
  include could not find load file:

    settings.cmake


-- The C compiler identification is GNU 9.4.0
-- Check for working C compiler: /usr/bin/gcc
-- Check for working C compiler: /usr/bin/gcc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- The CXX compiler identification is GNU 9.4.0
-- Check for working CXX compiler: /usr/bin/g++
-- Check for working CXX compiler: /usr/bin/g++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- The ASM compiler identification is GNU
-- Found assembler: /usr/bin/gcc
CMake Error at kernel/config.cmake:93 (include):
  include could not find load file:

    src/arch/KernelArchX86/config.cmake
Call Stack (most recent call first):
  kernel/CMakeLists.txt:41 (include)


CMake Error at kernel/CMakeLists.txt:141 (message):
  unknown KernelArch 'KernelArchX86'


-- Configuring incomplete, errors occurred!
See also "/home/mincheol/mnt2/awesome_system/build/CMakeFiles/CMakeOutput.log".

Can anyone help??

**Comment by axel-h** [comment]
`unknown KernelArch 'KernelArchX86'` is giving the right hint. Please check in the CMake files the values this variable can have.
The `settings.cmake`, is a file your system project is supposed to provide. This depends on the parts of the build systems you want to (re)use.
For a deeper analysis, I have to reproduce your setup first and check if the documentation there is still correct.

**Comment by mincheolsung** [comment]
@axel-h Thanks.
I wanted to run seL4 on a bare-metal machine and measure the IPC performance. However, I am quite confused and couldn't build it. With the interest of the time, I would like to run it on qemu-kvm. Can I start from sel4test-manifest?
Can you please point me to the proper tutorial?

**Comment by axel-h** [comment]
Yes, starting from https://github.com/seL4/sel4test-manifest (or in you case looking at https://github.com/seL4/sel4bench-manifest then for benchmarking) would be a better starting point. These are self-contained projects that can be built in the seL4 docker container out-of-the box. See https://docs.sel4.systems/GettingStarted and https://docs.sel4.systems/projects/sel4test for instructions. Also have a look at https://docs.sel4.systems/projects/dockerfiles about how to use the docker containers to avoid pitfalls setting up things on your host machine.
However, I'm not sure about what you plan to do on `qemu-kvm`. Usually QEMU is used as a full system emulator. The docker container has the QEMU binaries to do that.

**Comment by mincheolsung** [comment]
Thank you @axel-h. In the link (https://docs.sel4.systems/GettingStarted), I found an option for running seL4 on QEMU. My question is that does seL4 can be accelerated by KVM?

**Comment by axel-h** [comment]
> My question is that does seL4 can be accelerated by KVM?

I'm not aware there is special support in seL4 to run in on top of KVM. You should better ask this on the seL4 Mailinglist (https://lists.sel4.systems/postorius/lists/devel.sel4.systems/) or in the Forum at https://sel4.discourse.group with some more background what you plan to do.

**Comment by lsf37** [comment]
Just to confirm that there isn't any specific KVM support.

**Comment by mincheolsung** [comment]
Hi, I built sel4bench within the docker and booted it with a USB thumb drive.
I got a screen below:
<img width="722" alt="Screen Shot 2022-09-15 at 6 46 59 PM" src="https://user-images.githubusercontent.com/7972600/190521839-d4bd9e5a-1345-437e-855b-848f449d5ac4.png">
Is this expected? Does something go wrong?

**Comment by axel-h** [comment]
Sorry, at this point I'm a bit lost, because I'm not a x86 expert and not familiar with that output. Usually, you should see the ElfLoader output and then the kernel output first. This seem the native system now an not QEMU, I'm not sure if it works out of the box on any machine or if this needs some adaption first. Maybe somebody else can help here?

**Comment by Indanz** [comment]
I think you have to specify what UART to use somehow, otherwise seL4 doesn't know where to show output.

**Comment by mincheolsung** [comment]
I specified COM2 since I used supermicro's SOL (serial over lan) to communicate with the machine.
But still stuck on booting. Do I need to specify it somewhere else?

SERIAL **1** 115200
DEFAULT seL4test
LABEL seL4test
    kernel mboot.c32
    append sel4kernel --- rootserver

**Comment by mincheolsung** [comment]
Hi guys, eventually I was able to run sel4bench, yay! 
This is to update my status and I hope this can help someone with a similar issue.
As @Indanz said, I should have configured seL4 to use COM2 instead of COM1. I edited syslinux.cfg to `SERIAL 1 115200`. Also, I edited build/CMakeCache.txt:
`LibPlatSupportX86ConsoleDevice:STRING=com2`
`LibPlatSupportX86ConsoleDevice-ADVANCED:INTERNAL=2`
`LibPlatSupportX86ConsoleDeviceCom2:INTERNAL=ON`.
Then, I was able to see messages from the console.

<img width="981" alt="Screen Shot 2022-09-16 at 3 44 38 PM" src="https://user-images.githubusercontent.com/7972600/190719288-91a4854f-4a72-4457-8266-d93d321c6f91.png">

Now, I am exploring to benchmark the IPC of seL4. Thanks to all of you guys!
